File: algorithms/medium/number_of_good_leaf_nodes_pairs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countPairs(self, root: Optional[TreeNode], distance: int) -> int:
        self.ct = 0

        def dfs(node):
            if node.left is None and node.right is None:
                return [0]
            if node.left is None:
                return [n + 1 for n in dfs(node.right)]
            if node.right is None:
                return [n + 1 for n in dfs(node.left)]
            lft = [n + 1 for n in dfs(node.left)]
            rht = [n + 1 for n in dfs(node.right)]
            for l in lft:
                for r in rht:
                    if l + r <= distance:
                        self.ct += 1
            return lft + rht

        dfs(root)

        return self.ct

File: algorithms/medium/number_of_good_leaf_nodes_pairs.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

fn dfs(root: Option<Rc<RefCell<TreeNode>>>, distance: i32, ct: &mut i32) -> Vec<i32> {
    if root.is_none() {
        unreachable!();
    }

    let root = root.unwrap();
    let node = root.borrow();

    if node.left.is_none() && node.right.is_none() {
        return vec![0];
    }

    if node.left.is_none() {
        return dfs(node.right.clone(), distance, ct)
            .into_iter()
            .map(|n| n + 1)
            .collect();
    }

    if node.right.is_none() {
        return dfs(node.left.clone(), distance, ct)
            .into_iter()
            .map(|n| n + 1)
            .collect();
    }

    let mut lft = dfs(node.left.clone(), distance, ct)
        .into_iter()
        .map(|n| n + 1)
        .collect::<Vec<i32>>();
    let mut rht = dfs(node.right.clone(), distance, ct)
        .into_iter()
        .map(|n| n + 1)
        .collect::<Vec<i32>>();
    for l in &lft {
        for r in &rht {
            if l + r <= distance {
                *ct += 1;
            }
        }
    }

    lft.append(&mut rht);

    lft
}

impl Solution {
    pub fn count_pairs(root: Option<Rc<RefCell<TreeNode>>>, distance: i32) -> i32 {
        let mut ct = 0;
        dfs(root, distance, &mut ct);
        ct
    }
}
